### PR TITLE
fix for issue 19 network call failure snackbar

### PR DESCRIPTION
**Pull request by ppratikcr7**
This PR will alert user about which Network api calls failed as shown below:
https://user-images.githubusercontent.com/33730817/134628884-bf6dc121-40b0-4719-b70a-b5ed455cfa39.mov
Screenshot 2021-09-20 at 8 00 26 PM

### CODE REVIEW

File: frontend/projects/abtesting/src/app/core/http-interceptors/http-error.interceptor.ts [hunk 1]
```diff
@@ -4,25 +4,49 @@ import {
...
 @Injectable()
 export class HttpErrorInterceptor implements HttpInterceptor {
   constructor(
-    private authService: AuthService
+    private authService: AuthService,
+    private notificationsService: NotificationService
   ) { }
 
+  timeOut = 1500;
+  networkErrorCount: number = 0;
```
Comment (VivekFitkariwala) [inline frontend/projects/abtesting/src/app/core/http-interceptors/http-error.interceptor.ts:21]: Where are you decrementing the networkErrorCount?

File: frontend/projects/abtesting/src/app/core/http-interceptors/http-error.interceptor.ts [hunk 2]
```diff
@@ -4,25 +4,49 @@ import {
...
+  openSnackBar(error) {
+    let displayMessage = 'Network call failed for "' + error.url + '".';
+
+    if (!environment.production) {
+      displayMessage += ' See console for details.';
+    }
+    if (!((error as any).status === 401) && !environment.production) {
+      setTimeout(() => {
+        this.notificationsService.error(displayMessage); // this has 1500 as timeout
+      }, this.networkErrorCount * (this.timeOut+500)); // scheduling each snackbar one after the other with a gap of 500ms
```
Comment (VivekFitkariwala) [inline frontend/projects/abtesting/src/app/core/http-interceptors/http-error.interceptor.ts:36]: This seems incorrect. Correct me here but it should be
`(this.networkErrorCount * this.timeOut)+500`

File: frontend/projects/abtesting/src/app/core/http-interceptors/http-error.interceptor.ts (hunk 1 above)
Comment (ppratikcr7) [reply frontend/projects/abtesting/src/app/core/http-interceptors/http-error.interceptor.ts:21]: You are right, I am working on this.

File: frontend/projects/abtesting/src/app/core/http-interceptors/http-error.interceptor.ts (hunk 2 above)
Comment (ppratikcr7) [reply frontend/projects/abtesting/src/app/core/http-interceptors/http-error.interceptor.ts:36]: No, this is correct, if I do it as you mentioned the scheduling will be back to back without any gap of 500ms, only the first one would start after 500ms, rest would be exact next to first one.

Comment (VivekFitkariwala) [reply frontend/projects/abtesting/src/app/core/http-interceptors/http-error.interceptor.ts:36]: Yup got it!

**Comment by ppratikcr7** [comment]
Now we are using angular2-notifications library for network api call failure details throughout the app. Please find the below samples of error popups in both themes:
Dark theme: https://user-images.githubusercontent.com/33730817/136185382-5982f278-0b06-42e2-8ff7-eaf0b0b22204.mov
Light theme: https://user-images.githubusercontent.com/33730817/136185418-841ac5c6-b582-4ba6-b219-836b0bae8863.mov

File: frontend/projects/upgrade/src/app/core/http-interceptors/http-error.interceptor.ts [hunk 3]
```diff
@@ -4,25 +4,43 @@ import {
...
-      catchError(err => {
+      catchError( err => {
         if (err.status === 401) {
           // auto logout if 401 response returned from api
           this.authService.authLogout();
         }
-
-        return throwError(err);
+        this.openPopup(err);
+        return EMPTY; // returning EMPTY instead of throwError as Error is handled using snacker here itself
```
Comment (VivekFitkariwala) [inline frontend/projects/upgrade/src/app/core/http-interceptors/http-error.interceptor.ts:43]: Does this mean error not going to get printed on the console?

**Review by VivekFitkariwala — APPROVED** [review]

**Review by RidhamShah — APPROVED** [review]

**Comment by ppratikcr7** [comment]
@amurphy-cl We can merge this PR and close the corresponding issue.

**Pull request merged**